arm64 support

frozenprocess opened the pull request:
Adding arm64 support to the APIServer image.

## Description
In a multiarch setup apiserver fails to spawn when its assigned to an `arm64` node, this PR offers a solution to this problem.

```release-note
Add Calico API server support for arm64
```

caseydavenport reviewed (commented):
We should update the release process to build and push all images, and any multi-arch images as needed.

Reference points:


- https://github.com/projectcalico/apiserver/blob/c91ba80838e11520cc56147aaebef5d8d6aabe39/Makefile#L374
- https://github.com/projectcalico/apiserver/blob/c91ba80838e11520cc56147aaebef5d8d6aabe39/Makefile#L401

frozenprocess committed:
ARM64 support

Add: Enable `CGO` for arm64.
Add: Invoking `register` to add `binfmt_misc` support.
Modify: Append $(ARCH) suffix to `apiserver` and `filecheck` binaries.
Add: `--platform` to correct docker image tagging for `arm64`.
Add: `release-build` invokes image-all instead of image in order to support multiarch.o

HaveFun83 commented:
will be so great to see arm64 supported

caseydavenport commented on Makefile:
```diff
@@ -181,17 +181,17 @@ endif
 	mkdir -p $(CURDIR)/.lint-cache
 
-check-boring-ssl: $(BINDIR)/apiserver
+check-boring-ssl: $(BINDIR)/apiserver register
```
we only support this on amd64 so no point trying to assert this on non-amd architectures - they will fail.

caseydavenport commented on Makefile:
```diff
@@ -143,8 +143,8 @@ else
 	mkdir -p bin
-	$(DOCKER_RUN) -e CGO_ENABLED=$(CGO_ENABLED) $(CALICO_BUILD) \
-		sh -c '$(GIT_CONFIG_SSH) go build -v -i -o $@ -v $(LDFLAGS) $(PACKAGE_NAME)/cmd/apiserver'
+	$(DOCKER_RUN) -e CGO_ENABLED=$(CGO_ENABLED) $(CALICO_BUILD)-$(ARCH) \
```
Other repos don't use `$(CALICO_BUILD)-$(ARCH)` and instead just use `$(CALICO_BUILD)` - what's the reason for doing so here?

frozenprocess replied:
This is the binary build process and CGO_ENABLED is true.
`$(ARCH)` runs the build process inside an arm64 container.

We could omit that and use `CC` flag to cross compile the binary using `gcc` cross compiler.

caseydavenport commented:
The Calico team is reorganizing our code and repository structure in order to make development easier, and as such all of the code in this repository is now located at https://github.com/projectcalico/calico. We appreciate your PR and your patience as we work to improve our developer experience, and welcome you to re-open your changes against the new location. Please reach out if you have any questions!

zhangguanzhang commented:
is there any update?

zhangguanzhang commented:
https://github.com/projectcalico/calico/pull/5386
